### ISSUE TITLE
Rethink the interface to TreatmentPatterns

### ISSUE THREAD

**Issue by ablack3** [darwin-eu-dev/TreatmentPatterns#149]
Hi @edward-burn @catalamarti,

@mvankessel-EMC and I were discussing a new user interface for TreatmentPatterns. Maybe something like this..

```r
outputDataframe <- executeTreatmentPatterns(
    cdm,
    targetCohorts,
    eventCohorts,
    includeTreatments,
    periodPriorToIndex,
    minEraDuration,
    splitEventCohorts,
    splitTime,
    eraCollapseSize,
    combinationWindow,
    minPostCombinationDuration,
    filterTreatments,
    maxPathLength,
    minCellCount,
    minCellMethod,
    groupCombinations,
    addNoPaths) 
```

So the user would pass in the CDM, cohort information (dataframe or ids?), and all the execution settings and get a treatmentPatterns results object (or dataframe?) back which they could save where ever they wanted to in a separate step.

**Comment by edward-burn** [comment]
Yes, I think it would definitely be worth spending quite a bit of time thinking about how to best align the user interfaces for each of our off-the-shelf packages (so not just TreatmentPatterns, but also IncidencePrevalence, DrugUtilisation, and CohortSurvival). 

I personally think we can learn a lot from the tidyverse principles (https://design.tidyverse.org/index.html). Moreover, now we have quite a few users this is something where we should spend a lot of time with them to understand what they like/ dislike about current interfaces.

**Comment by catalamarti** [comment]
Happy to speak about that, it would be nice that all the packages have a similar way to work.
At first glance I think that there are some parameters that can be moved to the cohort creation step (minEraDuration, eraCollapseSize, ...) although I am not 100% sure. It would be nice to use a similar terminology across packages.

**Comment by mvankessel-EMC** [comment]
So I've been pruning the TreatmentPatterns objects a little. I'm planning to deprecate the SaveSettings object, with the transition to `Andromeda`, as all (intermeidate) results will be saved in that environment.

Besides that I'm planning rename `DataSettings` to `CDMSettings`, and give the option to either supply a `cdm` from CDMConnector, or `ConnectionDetails` with the required database schema.

I'm planning to do this in `R6`:

```
CDMSettings <- R6::R6Class(
  classname = "CDMSettings",
  inherit = Settings,
  public = list(
    initialize = function(cohortTable, connectionDetails = NULL, cdmDatabaseSchema = NULL, resultSchema = NULL, cdm = NULL) {
      private$connectionDetails <- connectionDetails
      private$cdmDatabaseSchema <- cdmDatabaseSchema
      private$resultSchema <- resultSchema
      private$cohortTable <- cohortTable
      private$cdm <- cdm
      
      self$validate()
      return(invisible(self))
    },
    
    validate = function() {
      errorMessages <- checkmate::makeAssertCollection()
      
      checkmate::assertClass(private$connectionDetails, "ConnectionDetails", null.ok = TRUE, add = errorMessages)
      checkmate::assertCharacter(x = private$connectionDetails$dbms, len = 1, null.ok = TRUE, add = errorMessages)
      checkmate::assertCharacter(private$cdmDatabaseSchema, null.ok = TRUE, len = 1, add = errorMessages)
      checkmate::assertCharacter(private$resultSchema, null.ok = TRUE, len = 1, add = errorMessages)
      checkmate::assertCharacter(private$cohortTable, null.ok = FALSE, len = 1, add = errorMessages)
      checkmate::assertClass(private$cdm, classes = "cdm_reference", null.ok = TRUE)
      checkmate::reportAssertions(collection = errorMessages)
      
      return(invisible(self))
    },
    
    fetchCohortTable = function(...) {
      if (is.null(private$cdm)) {
        extractCohortTable(...) # use TreatmentPatterns::extractCohortTable to fetch cohortTable
      } else if (is.null(private$connectionDetails)) {
          return(cdm[[private$cohortTable]]) # get cdm$cohortTable
      }
    }
  ),
  private = list(
    # private ----
    connectionDetails = NULL,
    cdmDatabaseSchema = "",
    resultSchema = "",
    cohortTable = "",
    cdm = NULL
  )
)
```

So the `fetchChortTable()` method either calls the existing implementation using `DatabaseConnector`, or it will just return the specified `cohortTable` from the `cdm` object. The actual call to the method would be made during `constructPathways()`

**Comment by mvankessel-EMC** [comment]
Addressed in #152

**Comment by ablack3** [comment]
Nice!
